From 0be0517c7773677c70b3b672e791eca61420690d Mon Sep 17 00:00:00 2001
From: Muhammad Aksar <muhammad.aksar@intel.com>
Date: Thu, 15 Dec 2022 11:45:48 +0530
Subject: [PATCH] Add
 talk/owt/patches/0032-Enable-Autofocus-for-back-camera.patch

Tracked-On: OAM-105058
Signed-off-by: Muhammad Aksar <muhammad.aksar@intel.com>
---
 scripts/prepare_dev.py                        |  1 +
 ...032-Enable-Autofocus-for-back-camera.patch | 58 +++++++++++++++++++
 2 files changed, 59 insertions(+)
 create mode 100644 talk/owt/patches/0032-Enable-Autofocus-for-back-camera.patch

diff --git a/scripts/prepare_dev.py b/scripts/prepare_dev.py
index 0cf5fbd..f42546d 100644
--- a/scripts/prepare_dev.py
+++ b/scripts/prepare_dev.py
@@ -58,6 +58,7 @@ patchList = [
     ('0029-JNI-pass-a-tcae-object-to-java-layer.patch', WEBRTC_PATH),
     ('0030-Update-Android-SDK-to-31-build.patch', BUILD_PATH),
     ('0031-Add-Native-SEI-Info.patch', WEBRTC_PATH),
+    ('0032-Enable-Autofocus-for-back-camera.patch', WEBRTC_PATH),
 ]
 
 def _patch(ignoreFailures=False):
diff --git a/talk/owt/patches/0032-Enable-Autofocus-for-back-camera.patch b/talk/owt/patches/0032-Enable-Autofocus-for-back-camera.patch
new file mode 100644
index 0000000..040a644
--- /dev/null
+++ b/talk/owt/patches/0032-Enable-Autofocus-for-back-camera.patch
@@ -0,0 +1,58 @@
+From 0bf0a9f7a6f03c25dcbb3bdc4a5fb7c03475f264 Mon Sep 17 00:00:00 2001
+From: Muhammad Aksar <muhammad.aksar@intel.com>
+Date: Sun, 11 Dec 2022 18:00:55 +0530
+Subject: [PATCH] Enable Autofocus for back camera
+
+Tracked-On: OAM-105058
+Signed-off-by: Muhammad Aksar <muhammad.aksar@intel.com>
+---
+ .../src/java/org/webrtc/Camera1Session.java    | 18 ++++++++++++++----
+ 1 file changed, 14 insertions(+), 4 deletions(-)
+
+diff --git a/sdk/android/src/java/org/webrtc/Camera1Session.java b/sdk/android/src/java/org/webrtc/Camera1Session.java
+index 39af59adb0..0a802d5aae 100644
+--- a/sdk/android/src/java/org/webrtc/Camera1Session.java
++++ b/sdk/android/src/java/org/webrtc/Camera1Session.java
+@@ -96,7 +96,7 @@ class Camera1Session implements CameraSession {
+       final android.hardware.Camera.Parameters parameters = camera.getParameters();
+       captureFormat = findClosestCaptureFormat(parameters, width, height, framerate);
+       final Size pictureSize = findClosestPictureSize(parameters, width, height);
+-      updateCameraParameters(camera, parameters, captureFormat, pictureSize, captureToTexture);
++      updateCameraParameters(camera, parameters, captureFormat, pictureSize, captureToTexture, info);
+     } catch (RuntimeException e) {
+       camera.release();
+       callback.onFailure(FailureType.ERROR, e.getMessage());
+@@ -120,7 +120,7 @@ class Camera1Session implements CameraSession {
+ 
+   private static void updateCameraParameters(android.hardware.Camera camera,
+       android.hardware.Camera.Parameters parameters, CaptureFormat captureFormat, Size pictureSize,
+-      boolean captureToTexture) {
++      boolean captureToTexture, android.hardware.Camera.CameraInfo info) {
+     final List<String> focusModes = parameters.getSupportedFocusModes();
+ 
+     parameters.setPreviewFpsRange(captureFormat.framerate.min, captureFormat.framerate.max);
+@@ -133,9 +133,19 @@ class Camera1Session implements CameraSession {
+     if (parameters.isVideoStabilizationSupported()) {
+       parameters.setVideoStabilization(true);
+     }
+-    if (focusModes.contains(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_VIDEO)) {
+-      parameters.setFocusMode(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_VIDEO);
++
++    // Enabling AF only for back camera since front camera would be normally based
++    // on fixed focus.
++    if (info.facing == android.hardware.Camera.CameraInfo.CAMERA_FACING_BACK) {
++        if (focusModes.contains(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE)) {
++          parameters.setFocusMode(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_PICTURE);
++          Logging.d(TAG, "Set CONTINUOUS_PICTURE AF for back camera");
++        } else if (focusModes.contains(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_VIDEO)) {
++          parameters.setFocusMode(android.hardware.Camera.Parameters.FOCUS_MODE_CONTINUOUS_VIDEO);
++          Logging.d(TAG, "Set CONTINUOUS_VIDEO AF for back camera");
++        }
+     }
++
+     camera.setParameters(parameters);
+   }
+ 
+-- 
+2.39.0
+
-- 
2.39.0

